powerdns_sync/lib/Atomia/DNS/PowerDNSDatabase.pm

```diff
@@ -417,4 +417,102 @@ sub remove_slave_zone {
 	}
 }
 
+sub add_tsig_key {
+	my $self = shift;
+	my $tsig_key_name = shift;
+	my $tsig_key_data = shift;
+
+	die "bad input data to add tsig key" unless defined($tsig_key_name) && ref($tsig_key_name) eq "" && $tsig_key_name =~ /^[a-zA-Z0-9_-]*$/ && defined($tsig_key_data) && ref($tsig_key_data) eq "HASH";
+	die "tsig secret missing" unless defined($tsig_key_data->{"secret"}) && length($tsig_key_data->{"secret"}) > 0;
+	die "algorithm missing" unless defined($tsig_key_data->{"algorithm"}) && length($tsig_key_data->{"algorithm"}) > 0;
+
+	eval {
+		my $name = $self->dbi->quote($tsig_key_name);
+		my $secret = $self->dbi->quote($tsig_key_data->{"secret"});
+		my $algorithm = $self->dbi->quote($tsig_key_data->{"algorithm"});
+
+		$self->dbi->do("DELETE FROM tsigkeys WHERE name = $name") || die "error removing previous tsigkey with the same name in add_tsig_key: $DBI::errstr";
+
+		my $query = "INSERT INTO tsigkeys (name, algorithm, secret) VALUES ($name, $algorithm, $secret)";
```

Comment:
If we delete the key before adding a new one with the same name and use the Account number to prevent adding the same key for the account we can use the console to add the same key and avoid the validation in the AutomationServer. 

@nemanja-zivkovic-atomia has tried to add reproduce this case but adding key failed on line 436. I'm not sure how the code has skipped line 434 but it should be checked. 

The same case can be tested for the function unassign_tsig_key above.

Reply:
I agree. @nemanja-zivkovic-atomia did you add this DELETE because this is what is done in other similar functions or you added it because you wanted to avoid duplicate key errors?

Reply:
In general, we should throw an error if you are trying to add the same key twice I think.

Reply:
> I'm not sure how the code has skipped line 434 but it should be checked.
What do you mean it failed on line 436? I think it didn't even reach that line or powerDNS code at all. It failed because of a constraint on 'tsigkey' table in AtomiaDNS database. So, as long as there is a constraint in place there: the delete operation won't do anything here.

Reply:
Aha, you think its failing here:
https://github.com/atomia/atomiadns/pull/60/files#diff-7c2fbca61c6200e9e96ca5dc3a1241bf1f8d50e367475c2fcc49c7c5a3ad93bdR15
because here we are setting it to be unique:
https://github.com/atomia/atomiadns/pull/60/files#diff-dfd0d70cd2f678974161db9bf01b05685b7d1f0d8be97f56c825a3fb399a3f75R347.
And I see that a similar thing is already done in the slave zone function:
https://github.com/atomia/atomiadns/blob/develop/powerdns_sync/lib/Atomia/DNS/PowerDNSDatabase.pm#L379